### ISSUE TITLE
Is this repo still mainted by developer？

### ISSUE THREAD

**Issue by spwpun** [openntpd-portable/openntpd-portable#73]
The same question for ntp(ntp.org).

**Comment by acts-1631** [comment]
Seems this repo is not maintained right now.

ntp.org is unrelated to OpenNTPD.

**Comment by arachsys** [comment]
If it's of interest to anyone, I have a somewhat different port of OpenBSD openntpd to Linux (glibc/musl) and the other BSDs at https://github.com/arachsys/openntpd. This is actively maintained and synced with OpenBSD at each release. (I guess I have another merge/sync coming up soon with OpenBSD 7.7.)

This was done for my own use and I don't feel obliged to follow upstream behaviour for compatibility where I can improve things, so it may not be to everyone's taste. (For example, I have replaced the messy ntpctl interface with a simpler ntpq command line syntax. However, the commit history is properly written with detailed discussion of these kinds of changes.) I also just use plain makefiles (compatible with BSD and GNU make) rather than the messy complexity of autotools, so target a somewhat more limited set of platforms.

There are some real bugs for non-OpenBSD platforms fixed in there that would be worth cherry picking, so if anyone decides to do their own port, please help yourself! I think at least

  https://github.com/arachsys/openntpd/commit/90f7ed0d
  https://github.com/arachsys/openntpd/commit/9979830e
  https://github.com/arachsys/openntpd/commit/616970fe
  https://github.com/arachsys/openntpd/commit/400c28b6
  https://github.com/arachsys/openntpd/commit/d8f53138
  https://github.com/arachsys/openntpd/commit/931555d6

are significant and worth fixing.

**Comment by acts-1631** [comment]
The main reason (I believe) people use OpenNTPD is because it's heavily audited by OpenBSD developers who have a great track record for security. I wouldn't personally want to use a fork by someone else, even if it's more up to date.

**Comment by arachsys** [comment]
Agreed, I offer the link more for people who would like to try their hand at their own fork rather than because I believe my own preferences will necessarily be a good fit for others. (Equally, you shouldn't assume this unmaintained -portable branch with loads of platform specific hacks has the same security profile as OpenBSD ntpd proper. This isn't like openssh-portable.)

I don't plan to advertise/market my branch for more general consumption because I don't have time to field interactions with potential users and contributors. I just needed a maintained ntpd with the bugs and annoyances fixed so I made myself one.

But building the upstream code instead of the -portable branch isn't half as much work as people might otherwise think, and you don't need to infest the project with autotools gloop or great complexity - the upstream code works pretty well with very minimal shims to platform-specific APIs, and a handful of bugs/assumptions fixed.

**Comment by busterb** [comment]
@arachsys alright if I take a look at your branch and pull in some of the fixes? I appreciate you maintaining a ~fork~ alternate port all this time.

**Comment by arachsys** [comment]
Of course, you're very welcome; please be my guest! There will be lots of changes you don't want to use because this version is for my own systems and I intentionally took the opportunity to reshape the interface and behaviour a little to suit my own taste.

But I think at least

https://github.com/arachsys/openntpd/commit/90f7ed0d76f3fb89b6533209807664019adb2da6
https://github.com/arachsys/openntpd/commit/9979830e321b282e2e93de6b0e4dd9c4041ffb35
https://github.com/arachsys/openntpd/commit/931555d68d1418e0b3fbd9464ca5831693dc0bd2
https://github.com/arachsys/openntpd/commit/d8f531380b15f8e8082684483446926dcef339d5
https://github.com/arachsys/openntpd/commit/616970fe88fe2280046288721d328e46311e68f8
https://github.com/arachsys/openntpd/commit/400c28b6c4e7b94d0f3870ce060606f8cce5fcb5

fix real problems.

The repo is structured as an upstream branch, which pulls in the upstream source files umodified, which is merged into a master branch which carries my changes. I update the upstream and remerge it back in each OpenBSD release, or if I see any changes in the upstream files that look wiser to merge sooner.

**Comment by arachsys** [comment]
Also worth checking

  https://github.com/arachsys/openntpd/blob/master/linux/adjfreq.c

maybe? You'll see I implement both adjfreq() and adjtime() in terms of the equivalent adjtimex() syscalls on Linux. The reason for this is that on linux-musl systems, adjtime() fails with EINVAL when called beyond the (limited) supported range, rather than just clamping into it, and this causes openntpd to behave in strange ways: 

 https://git.musl-libc.org/cgit/musl/tree/src/linux/adjtime.c

**Comment by busterb** [comment]
Thanks @arachsys, I'm testing this set of patches now:
https://github.com/openntpd-portable/openntpd-portable/tree/portable-fixes

**Comment by arachsys** [comment]
Great, thanks; glad they were useful! The fix for invalidating peer and sensor offsets when stepping the clock might also be a bug in the upstream OpenBSD code? (All the other issues aren't possible to hit on OpenBSD as far as I can tell, e.g. time_t is always 64-bit and monotonic clock isn't zero at boot.)

**Comment by perkelix** [comment]
Are we in sync now?

**Comment by busterb** [comment]
Yes, I agree @arachsys .  @perkelix we are not quite in sync; I think there are few more patches that could still go upstream. Will work on those in the coming release, since the OPENBSD_7_9 branch has been locked for some time.

**Comment by perkelix** [comment]
@busterb I think that we can close this and a few other bugs, now that 7.9p1 is released?

**Comment by busterb** [comment]
Sure.
